voip_utils/sip.py

```diff
@@ -190,87 +270,203 @@ def datagram_received(self, data: bytes, addr):
+                if re_uri is None:
+                    raise ValueError("Receiver URI did not match expected pattern")
+
+                server_ip = re_uri.group("host")
+                if not is_ipv4_address(server_ip):
+                    raise VoipError(f"Invalid IPv4 address in {ruri}")
+
+                self.on_call(
+                    CallInfo(
+                        caller_endpoint=caller_endpoint,
+                        caller_rtp_port=caller_rtp_port,
+                        server_ip=server_ip,
+                        headers=headers,
+                        opus_payload_type=opus_payload_type,
+                    )
+                )
+            elif method == "sip/2.0":
+                # Reply message means we must have received a response to someone we called
+                # TODO: Verify that the call / sequence IDs match our outgoing INVITE
+                _LOGGER.debug("Received INVITE response [%s]", message)
+                is_ok = False
+                protocol, code, reason, headers, body = self._parse_sip_reply(message)
+                if (code == "200") and (reason == "OK"):
+                    is_ok = True
+                if not is_ok:
+                    _LOGGER.debug("Received non-OK response [%s]", message)
+                    return
+
+                _LOGGER.debug("Got OK message")
+                if self.transport is None:
+                    _LOGGER.debug("No transport for exchanging SIP message")
+                    return
+                rtp_info = get_rtp_info(body)
+                remote_rtp_ip = rtp_info.rtp_ip
+                remote_rtp_port = rtp_info.rtp_port
+                opus_payload_type = rtp_info.payload_type
+                to_header = headers["to"]
+                caller_endpoint = None
+                if headers.get("to") is not None:
+                    caller_endpoint = SipEndpoint(headers.get("to", ""))
+                else:
+                    caller_endpoint = get_sip_endpoint(caller_ip, port=caller_sip_port)
+                # The From header should give us the URI used for sending SIP messages to the device
+                local_endpoint = None
+                if headers.get("from") is not None:
+                    local_endpoint = SipEndpoint(headers.get("from", ""))
+                else:
+                    local_endpoint = get_sip_endpoint(caller_ip, port=caller_sip_port)
+
+                _LOGGER.debug("Outgoing call to endpoint=%s", caller_endpoint)
+                call_id = headers["call-id"]
+                ack_lines = [
+                    f"ACK {caller_endpoint.uri} SIP/2.0",
+                    f"Via: SIP/2.0/UDP {local_endpoint.host}:{local_endpoint.port}",
+                    f"From: {local_endpoint.sip_header}",
+                    f"To: {to_header}",
+                    f"Call-ID: {call_id}",
+                    "CSeq: 50 ACK",
+                    "User-Agent: test-agent 1.0",
```

Comment:
Should we allow the consumer of this library to pass in a user agent ? Or should we put in `voip-utils` ?